modules/basegl/src/processors/shadercomponentprocessorbase.cpp

```diff
@@ -0,0 +1,138 @@
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+
+#include <modules/basegl/processors/shadercomponentprocessorbase.h>
+
+#include <inviwo/core/util/rendercontext.h>
+
+#include <modules/opengl/volume/volumegl.h>
+#include <modules/opengl/texture/textureunit.h>
+#include <modules/opengl/texture/textureutils.h>
+#include <modules/opengl/shader/shaderutils.h>
+#include <modules/opengl/shader/shadermanager.h>
+
+#include <modules/basegl/shadercomponents/shadercomponent.h>
+
+namespace inviwo {
+
+ShaderComponentProcessorBase::ShaderComponentProcessorBase(
+    const std::vector<std::pair<ShaderType, std::shared_ptr<const ShaderResource>>>&
+        shaderResources,
+    std::string_view identifier, std::string_view displayName)
+    : Processor(identifier, displayName)
+    , outport_("outport")
+    , shader_{shaderResources, Shader::Build::No}
+    , components_{} {
+
+    shader_.onReload([this]() { invalidate(InvalidationLevel::InvalidResources); });
+
+    addPort(outport_, "images");
+}
+
+ShaderComponentProcessorBase::~ShaderComponentProcessorBase() = default;
+
+void ShaderComponentProcessorBase::registerComponent(ShaderComponent& component) {
+    components_.push_back(&component);
+    for (const auto& [port, group] : component.getInports()) {
+        addPort(*port, group);
+    }
+    for (auto prop : component.getProperties()) {
+        addProperty(*prop);
+    }
+}
+
+void ShaderComponentProcessorBase::initializeResources() {
+    for (auto* comp : components_) {
+        try {
+            comp->initializeResources(shader_);
+        } catch (...) {
+            handleError("initializeResources", comp->getName());
+        }
+    }
+
+    auto* fso = shader_.getFragmentShaderObject();
+    fso->clearOutDeclarations();
```

Comment:
I assume that segments are currently only intended for fragment shaders?

Reply:
Yes all segments are for the fragment shader. Should be easy to do the same for other shader if the need arises...